### DISCUSSION TITLE
Training NER and TextCategorizer together

### DISCUSSION THREAD

**Discussion by darioprencipe** [explosion/spaCy#12991] [Help: Model Advice] [answered]
Hello,

I have a dataset of raw bank transactions. My task - out of a raw bank transaction description like `PAGAMENTO ADUE COD. DISP.:0122061307830975 NOME:GOOGLE IRELAND LIMITED - MANDATO:26941599`- is extracting the following info:

- counterpart >> `GOOGLE IRELAND LIMITED` (NER task)
- reason >>  `MANDATO:26941599` (NER task)
- type >> `DIRECT_DEBIT` (textcat task)

I have already trained from scratch a model with `ner` (and `tok2vec`) component to extract counterpart entity. Setup of the training and test datasets in `.spacy` format and training task using spacy CLI and `.cfg` file was pretty straightforward. 

What I don't get now is - provided I do have the raw annotations (let's say in a .csv file) - how I should go about training `ner` AND `textcat` **together** in the same blank sheet model? Shall I:
1. build single train.spacy and dev.spacy files using a blank sheet model, where I do assign both entities and categories at the same time, and then running one single training task using .cfg to declare which components I want to train?
2. OR build distinct train/test dataset couples and run 2 distinct training tasks assembling the models together afterward?

The output I imagine is a custom spacy model that allows me to do the following:

```
import spacy

nlp = spacy.load("training/model-best")

doc = nlp("PAGAMENTO ADUE COD. DISP.:0122061307830975 NOME:GOOGLE IRELAND LIMITED - MANDATO:26941599")

doc.ents
# prints me the 2 entities ("COUNTERPART": "GOOGLE IRELAND LIMITED", "REASON": "MANDATO:26941599")

doc.cats
# prints me the estimated category => DIRECT_DEBIT
```

In general, I couldn't find examples of pipeline where you or folks around do exactly this. So I wonder what's the best-practice-way to approach this problem.

Thanks a lot!
Dario

**Comment by danieldk** [comment, marked as answer]
If you have both the entity annotations and the text category in the same data (which you seem to have), then it is best to treat this as as single training task. This will allow both the named entity recognizer and text categorizer to use the same underlying contextual vector (tok2vec representations).

Training two separate pipelines and merging them is more of a last resort when the data sets are disjoint. E.g. if you had one document collection with text categories and a completely different document collection with named entity annotations.

**Reply by darioprencipe** [reply]
Thanks a lot @danieldk ! I have one comment and 2 questions:
- **comment**: I believe this use case is pretty common! I think it could be beneficial to add some documentation on how different components should be trained together. 
- **question**: assume I have everything in the same .csv. At some point I will have to convert this into .spacy format. Is there any specific data format I should comply with in order to be able to pass both entities and categories in the same conversion loop?
- **question**: is there any config you would recommend for training together a NER and textCat on a dataset of semantically-poor text records? The .cfg file I've been using for the NER task so far (with good results) is the following:

```
[paths]
train = null
dev = null
vectors = "it_core_news_lg"
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "it"
pipeline = ["tok2vec","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
scorer = {"@scorers":"spacy.ner_scorer.v1"}
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,1000,2500,2500]
include_static_vectors = true

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 256
depth = 8
window_size = 1
maxout_pieces = 3

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1200
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null
before_update = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
ents_f = 1.0
ents_p = 0.0
ents_r = 0.0
ents_per_type = null

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

Thanks!!

**Reply by danieldk** [reply]
> question: assume I have everything in the same .csv. At some point I will have to convert this into .spacy format. Is there any specific data format I should comply with in order to be able to pass both entities and categories in the same conversion loop?

No, you can just create `Doc`s with the contents and set the NER and category annotations.

> question: is there any config you would recommend for training together a NER and textCat on a dataset of semantically-poor text records? The .cfg file I've been using for the NER task so far (with good results) is the following:

If you have a very large number of records (they do not need to be annotated), you could consider using `spacy pretrain` to pretrain domain-specific vectors: https://spacy.io/usage/embeddings-transformers/#pretraining

I think you could also get some mileage out of inspecting the output of the tokenizer and seeing if you could add some specific rules to improve it. In the example that you provided the tokens were:

```
In [6]: [token.text for token in doc]
Out[6]: 
['PAGAMENTO',
 'ADUE',
 'COD',
 '.',
 'DISP.:0122061307830975',
 'NOME',
 ':',
 'GOOGLE',
 'IRELAND',
 'LIMITED',
 '-',
 'MANDATO:26941599'
```

I am not familiar with the data/domain, but I could imagine that if e.g. _ MANDATO_ is a good indicator for class, then the tokenization here would not be optimal. The data could contain multiple _ MANDATO:<somenumber>_ and you'd run into data sparseness issues. Whereas if this was tokenized into e.g. __ MANDATO : 26941599_, the model could pick up _MANDATO_ as a good feature across documents.

**Reply by darioprencipe** [reply]
@danieldk thanks a lot - that definitely helps. I don't get whether I fully understood your point on tokens => 

> I could imagine that if e.g. _ MANDATO_ is a good indicator for class

Correct, it is. While indeed tokenizer currently doesn't split that >> `'MANDATO:26941599'`. However, then you say:

> Whereas if this was tokenized into e.g. __ MANDATO : 26941599_, the model could pick up MANDATO as a good feature across documents.

Isn't this exactly how spaCy tokenized it? Following your point I would have expected the desired token here was `MANDATO:`, as it's a common, repeated pattern in the documents and it indeed helps predicting a category label.

**Reply by danieldk** [reply]
Sorry for the unclarity. By convention I put spaces between the tokens in `MANDATO : 26941599` to signify that in a better tokenization they'd be separate tokens.

**Reply by darioprencipe** [reply]
@danieldk sorry to bother again, just wondering - how and when should I amend the tokenizer to include, e.g. rule that always splits tokens whenever ":" is found, if I do pre-training on my own corpus?

Shall I do it right before building the annotated `train.spacy` and `dev.spacy` files, as I start from a blank `nlp` object?  

Thanks!

**Reply by rmitsch** [reply]
Your corpus should be tokenized consistently, i. e. training and inference data should use the same tokenizer.
